devicetypes/smartthings/matrix-zdb5100.src/matrix-zdb5100.groovy

```diff
@@ -0,0 +1,1272 @@
+}
+
+/**
+ *  zwaveEvent( COMMAND_CLASS_VERSION V1 (0x86) : VERSION_REPORT )
+ *
+ *  The Version Report Command is used to advertise the library type, protocol version, and application version.
+
+ *  Action: Publish values as device 'data' and log an info message. No check is performed.
+ *
+ **/
+def zwaveEvent(physicalgraph.zwave.commands.versionv1.VersionReport cmd) {
+    log.debug("zwaveEvent(): Version Report received: ${cmd}")
+
+    def zWaveLibraryTypeDisp  = String.format("%02X", cmd.zWaveLibraryType)
+
+    def applicationVersionDisp = String.format("%d.%02d", cmd.applicationVersion, cmd.applicationSubVersion)
+    def zWaveProtocolVersionDisp = String.format("%d.%02d", cmd.zWaveProtocolVersion, cmd.zWaveProtocolSubVersion)
+
+    log.debug("Version Report: Application Version: ${applicationVersionDisp}, " +
+              "Z-Wave Protocol Version: ${zWaveProtocolVersionDisp}, " +
+              "Z-Wave Library Type: ${zWaveLibraryTypeDisp}")
+
+    updateDataValue("applicationVersion", "${cmd.applicationVersion}")
+    updateDataValue("applicationSubVersion", "${cmd.applicationSubVersion}")
+    updateDataValue("zWaveLibraryType", "${zWaveLibraryTypeDisp}")
+    updateDataValue("zWaveProtocolVersion", "${cmd.zWaveProtocolVersion}")
+    updateDataValue("zWaveProtocolSubVersion", "${cmd.zWaveProtocolSubVersion}")
+}
+
+/**
+ *  dimmerEvent()
+ *
+ *  Common handler for BasicReport, SwitchBinaryReport, SwitchMultilevelReport.
+ *
+ *  Action: Raise 'switch' and 'level' events.
+ *   Restore pending level if dimmer has been switched on after nightmode has been disabled.
+ *   
+ **/
+def dimmerEvent(physicalgraph.zwave.Command cmd) {
+    def result = []
+
+    // switch event:
+    def switchValue = (cmd.value ? "on" : "off")
+
+    def switchEvent = createEvent(name: "switch", value: switchValue)
+
+    if (switchEvent.isStateChange) log.debug("Dimmer turned ${switchValue}.")
+    result << switchEvent
+
+
+    // level event:
+    def levelValue = Math.round(cmd.value * 100 / 99)
+    if (levelValue != 0 && levelValue != 1) {
+        state.lastDimmerValue = levelValue
+    }
+    if (levelValue == 1) {
+        levelValue = state.DIM_OFF_OFFSET
+    }
+
+    def levelEvent = createEvent(name: "level", value: levelValue, unit: "%") //+state.DIM_OFF_OFFSET
+    if (levelEvent.isStateChange) log.debug("Dimmer level is ${levelValue}%")
+    result << levelEvent
+
+    return result
+}
+
+
+/*****************************************************************************************************************
+ *  Capability-related Commands:
+ *****************************************************************************************************************/
+
+/**
+ *  on()                        [Capability: Switch]
+ *
+ *  Turn the dimmer on.
+ **/
+def on() {
+    def result = []
+    result << endpointCmd(zwave.basicV1.basicSet(value: 255), 5).format()
```

Comment:
Again, why don't you just use something like:
```
def result = []

if (state?.DIM_OFF_OFFSET > 0) {
	result << endpointCmd(zwave.switchMultilevelV3.switchMultilevelSet(value: state.lastDimmerValue), 5).format()
} else {
	result << endpointCmd(zwave.basicV1.basicSet(value: 255), 5).format()
}
```
It's way more elegant

Reply:
Yes again I must agree. Updated.